xgi/algorithms/assortativity.py

```diff
@@ -0,0 +1,156 @@
+import random
+from itertools import combinations
+
+import numpy as np
+
+import xgi
+from xgi.exception import XGIError
+
+__all__ = ["dynamical_assortativity", "degree_assortativity"]
+
+
+def dynamical_assortativity(H):
+    """Computes the dynamical assortativity of a uniform hypergraph.
+
+    Parameters
+    ----------
+    H : xgi.Hypergraph
+        Hypergraph of interest
+
+    Returns
+    -------
+    float
+        The dynamical assortativity
+
+    Raises
+    ------
+    XGIError
+        If the hypergraph is not uniform, or if there are no nodes
+        or no edges
+
+    References
+    ----------
+    Nicholas Landry and Juan G. Restrepo,
+    Hypergraph assortativity: A dynamical systems perspective,
+    Chaos 2022.
+    DOI: 10.1063/5.0086905
+
+    """
+    if not xgi.is_uniform(H):
+        raise XGIError("Hypergraph must be uniform!")
+
+    if H.num_nodes == 0 or H.num_edges == 0:
+        raise XGIError("Hypergraph must contain nodes and edges!")
+
+    degs = H.degree()
+    k1 = np.mean([degs[n] for n in H.nodes])
+    k2 = np.mean([degs[n] ** 2 for n in H.nodes])
```

Comment:
Nitpick: this makes two passes over the set of nodes, but one is possible, something along these lines
```suggestion
    k1, k2 = np.array([(degs[n], degs[n]**2) for n in H.nodes]).mean(axis=1)
```

Not sure about the `axis=1`, but you get the idea.

Reply:
Good point. I changed this for a 1.5x speedup for k1 and a 4.5x speedup for k2. I think the list comprehension was the slowest part so that's what I removed.